File: bindings/typescript/src/safe_monotonic.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeMonotonic - Typed wrapper for monotonically increasing sequences.
 *
 * Delegates all computation to the JavaScript FFI binding, which calls
 * libproven (Idris 2 + Zig) via Deno.dlopen. No logic is reimplemented here.
 *
 * @module
 */

import {
  MonotonicCounter as JsMonotonicCounter,
  MonotonicTimestamp as JsMonotonicTimestamp,
  MonotonicId as JsMonotonicId,
  MonotonicSequence as JsMonotonicSequence,
  SafeMonotonic as JsSafeMonotonic,
} from '../../javascript/src/safe_monotonic.js';

/** Result type returned by monotonic operations. */
export type Result<T> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly error: string };

/**
 * Typed wrapper around the FFI-backed MonotonicCounter.
 *
 * Every method calls through to the JavaScript FFI wrapper, which in turn
 * calls the formally verified Idris 2 code via the Zig FFI bridge.
 */
export class MonotonicCounter {
  /** The underlying JS FFI monotonic counter instance. */
  readonly #inner: InstanceType<typeof JsMonotonicCounter>;

  /** Private constructor -- use MonotonicCounter.create() instead. */
  private constructor(inner: InstanceType<typeof JsMonotonicCounter>) {
    this.#inner = inner;
  }

  /**
   * Create a monotonic counter via FFI.
   * Delegates to proven_monotonic_create.
   *
   * @param initial - Initial value.
   * @param maxValue - Maximum value before wrapping.
   * @returns Result with the MonotonicCounter instance, or error.
   */
  static create(initial: number | bigint, maxValue: number | bigint): Result<MonotonicCounter> {
    const result = JsSafeMonotonic.counter(
      initial,
      maxValue,
    ) as Result<InstanceType<typeof JsMonotonicCounter>>;
    if (!result.ok) return result;
    return { ok: true, value: new MonotonicCounter(result.value) };
  }

  /**
   * Get the next value (atomically increments).
   * Delegates to proven_monotonic_next via FFI.
   *
   * @returns Result with the next counter value, or error.
   */
  next(): Result<number> {
    return this.#inner.next() as Result<number>;
  }

  /**
   * Close and free the native counter.
   * Delegates to proven_monotonic_free via FFI.
   */
  close(): void {
    this.#inner.close();
  }
}

/**
 * Monotonic timestamp (placeholder).
 * Will wrap a counter with clock source when the Idris implementation
 * is complete.
 */
export class MonotonicTimestamp extends JsMonotonicTimestamp {}

/**
 * Monotonic ID generator (placeholder).
 * Will be wired when the Idris implementation is complete.
 */
export class MonotonicId extends JsMonotonicId {}

/**
 * Monotonic sequence (placeholder).
 * Will be wired when the Idris implementation is complete.
 */
export class MonotonicSequence extends JsMonotonicSequence {}

/**
 * SafeMonotonic namespace.
 * All operations delegate to the JavaScript FFI binding.
 */
export class SafeMonotonic {
  /**
   * Create a monotonic counter.
   * Delegates to proven_monotonic_create via FFI.
   *
   * @param initial - Initial value.
   * @param maxValue - Maximum value.
   * @returns Result with the MonotonicCounter instance, or error.
   */
  static counter(initial: number | bigint, maxValue: number | bigint): Result<MonotonicCounter> {
    return MonotonicCounter.create(initial, maxValue);
  }
}

File: bindings/ruby/lib/proven/safe_monotonic.rb
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
# frozen_string_literal: true

# Safe monotonic counter operations via FFI to libproven.
# ALL computation delegates to Idris 2 compiled code.
# Monotonic counters are opaque pointers managed by libproven.

module Proven
  module SafeMonotonic
    class << self
      # Create a monotonic counter.
      # Returns an opaque handle or nil on error.
      # Caller MUST call #free when done.
      #
      # @param initial [Integer] initial value
      # @param max_value [Integer] maximum value before wrapping
      # @return [Fiddle::Pointer, nil]
      def create(initial, max_value)
        FFI.invoke_ptr(
          "proven_monotonic_create",
          [Fiddle::TYPE_LONG_LONG, Fiddle::TYPE_LONG_LONG],
          [initial, max_value]
        )
      end

      # Get the next value from the counter.
      # Returns nil if counter has reached max_value or error.
      #
      # @param counter [Fiddle::Pointer]
      # @return [Integer, nil]
      def next_value(counter)
        return nil if counter.nil?
        result = FFI.invoke_int_result(
          "proven_monotonic_next",
          [Fiddle::TYPE_VOIDP],
          [counter]
        )
        return nil unless result
        status, value = result
        status == FFI::STATUS_OK ? value : nil
      end

      # Free a monotonic counter handle.
      #
      # @param counter [Fiddle::Pointer]
      def free(counter)
        return if counter.nil?
        FFI.invoke_void(
          "proven_monotonic_free",
          [Fiddle::TYPE_VOIDP],
          [counter]
        )
      end
    end
  end
end
